library/core/src/num/dec2flt/decimal.rs

```diff
@@ -0,0 +1,384 @@
+            let quotient = n / 10;
+            let remainder = n - (10 * quotient);
+            if write_index < Self::MAX_DIGITS {
+                self.digits[write_index] = remainder as u8;
+            } else if remainder > 0 {
+                self.truncated = true;
+            }
+            n = quotient;
+        }
+        while n > 0 {
+            write_index -= 1;
+            let quotient = n / 10;
+            let remainder = n - (10 * quotient);
+            if write_index < Self::MAX_DIGITS {
+                self.digits[write_index] = remainder as u8;
+            } else if remainder > 0 {
+                self.truncated = true;
+            }
+            n = quotient;
+        }
+        self.num_digits += num_new_digits;
+        if self.num_digits > Self::MAX_DIGITS {
+            self.num_digits = Self::MAX_DIGITS;
+        }
+        self.decimal_point += num_new_digits as i32;
+        self.trim();
+    }
+
+    /// Computes decimal * 2^-shift.
+    pub fn right_shift(&mut self, shift: usize) {
+        let mut read_index = 0;
+        let mut write_index = 0;
+        let mut n = 0_u64;
+        while (n >> shift) == 0 {
+            if read_index < self.num_digits {
+                n = (10 * n) + self.digits[read_index] as u64;
+                read_index += 1;
+            } else if n == 0 {
+                return;
+            } else {
+                while (n >> shift) == 0 {
+                    n *= 10;
+                    read_index += 1;
+                }
+                break;
+            }
+        }
+        self.decimal_point -= read_index as i32 - 1;
+        if self.decimal_point < -Self::DECIMAL_POINT_RANGE {
+            self.num_digits = 0;
+            self.decimal_point = 0;
+            self.negative = false;
+            self.truncated = false;
+            return;
+        }
+        let mask = (1_u64 << shift) - 1;
+        while read_index < self.num_digits {
+            let new_digit = (n >> shift) as u8;
+            n = (10 * (n & mask)) + self.digits[read_index] as u64;
+            read_index += 1;
+            self.digits[write_index] = new_digit;
+            write_index += 1;
+        }
+        while n > 0 {
+            let new_digit = (n >> shift) as u8;
+            n = 10 * (n & mask);
+            if write_index < Self::MAX_DIGITS {
+                self.digits[write_index] = new_digit;
+                write_index += 1;
+            } else if new_digit > 0 {
+                self.truncated = true;
+            }
+        }
+        self.num_digits = write_index;
+        self.trim();
+    }
+}
+
+/// Parse a big integer representation of the float as a decimal.
+pub fn parse_decimal(mut s: &[u8]) -> Decimal {
+    // can't fail since it follows a call to parse_number
+    let mut d = Decimal::default();
+    let start = s;
+    // SAFETY: cannot be empty, must be called after check for empty
+    // input in parse_number.
+    let c = unsafe { s.first_unchecked() };
```

Comment:
If this has a non-local safety-invarient, please make it an `unsafe fn`.
```suggestion
pub unsafe fn parse_decimal(mut s: &[u8]) -> Decimal {
    // can't fail since it follows a call to parse_number
    let mut d = Decimal::default();
    let start = s;
    // SAFETY: cannot be empty, must be called after check for empty
    // input in parse_number.
    let c = unsafe { s.first_unchecked() };
```

Reply:
I think I'll probably see if I can make the initial steps outside for safety guarantees, and see if I can set the sign afterwards. This function shouldn't be unsafe inherently, but this comment is entirely true.

Reply:
[Fixed](https://github.com/Alexhuszagh/rust/blob/fb1a817fa13f774ce4da27fb9769db2779163030/library/core/src/num/dec2flt/decimal.rs#L204-L209) by removing `negative` from `Decimal` (which was never used, anyway) and moving the sign parsing logic into `dec2flt`. This means we can simply skip this unsafe functionality.

```rust
/// Parse a big integer representation of the float as a decimal.
pub fn parse_decimal(mut s: &[u8]) -> Decimal {
    let mut d = Decimal::default();
    let start = s;
    s = s.skip_chars(b'0');
    parse_digits(&mut s, |digit| d.try_add_digit(digit));
```